include/shark/Algorithms/DirectSearch/Operators/Hypervolume/HypervolumeContribution2D.h

```diff
@@ -0,0 +1,173 @@
+ *
+ *
+ * \par Copyright 1995-2016 Shark Development Team
+ * 
+ * <BR><HR>
+ * This file is part of Shark.
+ * <http://image.diku.dk/shark/>
+ * 
+ * Shark is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published 
+ * by the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ * 
+ * Shark is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ * 
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Shark.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+#ifndef SHARK_ALGORITHMS_DIRECTSEARCH_HYPERVOLUME_CONTRIBUTION_2D_H
+#define SHARK_ALGORITHMS_DIRECTSEARCH_HYPERVOLUME_CONTRIBUTION_2D_H
+
+#include <shark/LinAlg/Base.h>
+#include <shark/Core/utility/KeyValuePair.h>
+
+#include <algorithm>
+#include <vector>
+
+namespace shark {
+/// \brief Finds the least/largest Contributors given 2D points
+///
+/// The algorithm sorts the points by their first coordinate and then 
+/// calculates the volume of the hypervolume dominated by each point separately
+/// returning the index of the elements with the minimum/maximum contribution.
+struct HypervolumeContribution2D {
+private:
+	struct Point{
+		Point(){}
+
+		Point(double f1, double f2, std::size_t index)
+		: f1(f1)
+		, f2(f2)
+		, index(index)
+		{}
+			
+		bool operator<(Point const& rhs) const{//for lexicographic sorting
+			if (f1 < rhs.f1) return true;
+			if (f1 > rhs.f1) return false;
+			return (f2 < rhs.f1);
+		}
+		
+		double f1;
+		double f2;
+		std::size_t index;
+	};
+	
+	///\brief Stores the k elements with the best contribution(highest,lowest) as indicated by comparator
+	///
+	/// The algorithm returns the indizes of the front front[i].index in order of contribution.
+	/// The input is a set of points sorted by x-value. the edge-points are never selected.
+	///
+	/// This is implemented by using a min-heap that stores the k best elements,
+	/// but having the smallest element on top so that we can quickly decide which
+	/// element to remove.
+	template<class Comparator>
+	std::vector<std::size_t> bestContributors( std::vector<Point> const& front, std::size_t k, Comparator comp)const{
+		
+		std::vector<std::pair<double,std::size_t> > bestK(k+1);
+		auto heapStart = bestK.begin();
+		auto heapEnd = bestK.begin();
+		
+		auto pointComp = [&](std::pair<double,std::size_t> const& lhs, std::pair<double,std::size_t> const& rhs){return comp(lhs.first,rhs.first);};
+		
+		//compute the hypervalue contribution for each point except the endpoints
+		*heapEnd = std::make_pair((front[2].f1 - front[1].f1) * (front[0].f2 - front[1].f2),front[1].index);
+		++heapEnd;
+		for(std::size_t i = 2; i < front.size()-1;++i){
```

Comment:
no, the point i=1 is already inserted in line 82. hm maybe i can unify this...yeah probably